src/main/java/com/es/phoneshop/web/CheckoutPageServlet.java

```diff
@@ -0,0 +1,103 @@
+        var cart = cartService.getCart(request);
+        request.setAttribute("order", orderService.getOrder(cart));
+        request.setAttribute("paymentMethods", orderService.getPaymentMethods());
+        request.getRequestDispatcher(CHECKOUT_JSP).forward(request, response);
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        var cart = cartService.getCart(request);
+        var order = orderService.getOrder(cart);
+        Map<String, String> errors = new HashMap<>();
+
+        setRequiredParameter(request, "firstName", errors, order::setFirstName);
+        setRequiredParameter(request, "lastName", errors, order::setLastName);
+        setRequiredParameter(request, "phone", errors, order::setPhone);
+        setRequiredParameter(request, "deliveryAddress", errors, order::setDeliveryAddress);
+        setDeliveryDate(request, errors, order);
+        setPaymentMethod(request, errors, order);
+
+        handleError(request, response, errors, order, cart);
+    }
+
+    private void handleError(HttpServletRequest request, HttpServletResponse response,
+                             Map<String, String> errors, Order order, Cart cart) throws IOException, ServletException {
+        if (errors.isEmpty()) {
+            orderService.placeOrder(order);
+            cartService.cleanCart(cart);
+            response.sendRedirect(request.getContextPath() + "/order/overview/" + order.getSecureId());
+        } else {
+            request.setAttribute("errors", errors);
+            request.setAttribute("order", order);
+            request.setAttribute("paymentMethods", orderService.getPaymentMethods());
+            request.getRequestDispatcher(CHECKOUT_JSP).forward(request, response);
+        }
+    }
+
+    private void setRequiredParameter(HttpServletRequest request, String parameter,
+                                      Map<String, String> errors, Consumer<String> consumer) {
+        var value = request.getParameter(parameter);
+        if (value == null || value.isEmpty()) {
+            errors.put(parameter, "Value is required");
+        } else {
+            consumer.accept(value);
+        }
+    }
+
+    private void setPaymentMethod(HttpServletRequest request, Map<String, String> errors, Order order) {
+        var value = request.getParameter("paymentMethod");
+        if (value == null || value.isEmpty()) {
+            errors.put("paymentMethod", "Value is required");
+        } else {
+            order.setPaymentMethod(PaymentMethod.valueOf(value));
+        }
+    }
+
+    private void setDeliveryDate(HttpServletRequest request, Map<String, String> errors, Order order) {
+        var value = request.getParameter("deliveryDate");
+        if (value.isEmpty()) {
+            errors.put("deliveryDate", "Value is required");
+        } else if (LocalDate.now().isAfter(LocalDate.parse(value))) {
```

Comment:
Add parse exception processing

Reply:
I use "date" type for this input and it has all checks